File: 022_MinStack.cpp
// Topic: Min Stack
// Description: Design stack supporting push, pop, top, and retrieving current minimum in O(1).
// Approach: Maintain main stack and stack of current minimums (or store paired values).

#include <bits/stdc++.h>
using namespace std;

class MinStack
{
    vector<pair<int, int>> st;  // {value, current_min}
   public:
    MinStack() {}
    void push(int x)
    {
        if (st.empty())
            st.emplace_back(x, x);
        else
            st.emplace_back(x, min(x, st.back().second));
    }
    void pop()
    {
        if (!st.empty())
            st.pop_back();
    }
    int top() const { return st.back().first; }
    int getMin() const { return st.back().second; }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    // Test scenario 1
    {
        MinStack ms;
        ms.push(-2);
        ms.push(0);
        ms.push(-3);
        cout << "Test 1: getMin -> " << ms.getMin() << "\n";  // expect -3
        ms.pop();
        cout << "Test 1: top -> " << ms.top() << "\n";        // expect 0
        cout << "Test 1: getMin -> " << ms.getMin() << "\n";  // expect -2
    }
    // Test scenario 2
    {
        MinStack ms;
        ms.push(2);
        ms.push(2);
        ms.pop();
        cout << "Test 2: top -> " << ms.top() << "\n";        // expect 2
        cout << "Test 2: getMin -> " << ms.getMin() << "\n";  // expect 2
    }
}

/*
Time Complexity: O(1) per operation
Space Complexity: O(n)
*/

File: 022_MinStack.py
# Topic: Min Stack
# Description: Design stack supporting push, pop, top, and retrieving current minimum in O(1).
# Approach: Maintain main stack and stack of current minimums (or store paired values).


class MinStack:
    def __init__(self):
        self.stack = []  # stores (value, current_min)

    def push(self, val: int) -> None:
        if not self.stack:
            self.stack.append((val, val))
        else:
            self.stack.append((val, min(val, self.stack[-1][1])))

    def pop(self) -> None:
        if self.stack:
            self.stack.pop()

    def top(self) -> int:
        return self.stack[-1][0]

    def getMin(self) -> int:
        return self.stack[-1][1]


if __name__ == "__main__":
    # Test scenario 1
    ms = MinStack()
    ms.push(-2)
    ms.push(0)
    ms.push(-3)
    print(f"Test 1: getMin -> {ms.getMin()}")  # expect -3
    ms.pop()
    print(f"Test 1: top -> {ms.top()}")  # expect 0
    print(f"Test 1: getMin -> {ms.getMin()}")  # expect -2

    # Test scenario 2
    ms2 = MinStack()
    ms2.push(2)
    ms2.push(2)
    ms2.pop()
    print(f"Test 2: top -> {ms2.top()}")  # expect 2
    print(f"Test 2: getMin -> {ms2.getMin()}")  # expect 2

"""
Time Complexity: O(1) per operation
Space Complexity: O(n)
"""
